File: iosApp/iosApp/DesignSystem/Components/ProductRow.swift
import SwiftUI
import Shared

/// Visual urgency bucket. The cuts live ONLY in `expirationState(_:)` — the
/// single source of truth reused by the detail section grouping, the day-number
/// colour, and the progress bar. (Mirror of Android `expirationStateOf`.)
enum ExpirationState {
    case bad, warn, fresh

    var color: Color {
        switch self {
        case .bad:   return .fridgeyRust
        case .warn:  return .fridgeyAmber
        case .fresh: return .fridgeyMintDeep
        }
    }
}

/// Design cuts: <=1 día → BAD, 2..7 → WARN, >7 → FRESH.
func expirationState(_ diasRestantes: Int) -> ExpirationState {
    if diasRestantes <= 1 { return .bad }
    if diasRestantes <= 7 { return .warn }
    return .fresh
}

/// Per-category tint for the emoji bubble (design palette); unlisted → smoke.
private func categoryIconBg(_ categoria: Categoria) -> Color {
    switch categoria {
    case .lacteos:  return Color(red: 0.933, green: 0.949, blue: 0.965)  // dairy  #EEF2F6
    case .carnes:   return Color(red: 0.961, green: 0.898, blue: 0.886)  // meat   #F5E5E2
    case .verduras: return Color(red: 0.898, green: 0.937, blue: 0.878)  // veg    #E5EFE0
    case .frutas:   return Color(red: 0.965, green: 0.922, blue: 0.863)  // fruit  #F6EBDC
    case .bebidas:  return Color(red: 0.910, green: 0.937, blue: 0.945)  // drink  #E8EFF1
    default:        return .fridgeySmoke
    }
}

/// One product inside a "balda" (parity with Android `ProductRow`):
/// 44 emoji bubble (tinted per category) · name + "Categoría · cantidad+unidad"
/// · big serif day count + label, with a 2-pt progress bar that widens as
/// expiry nears. Urgency is computed from `daysRemaining` via `expirationState`.
struct ProductRow: View {
    let categoria: Categoria
    let name: String
    let supporting: String
    let daysRemaining: Int

    private var state: ExpirationState { expirationState(daysRemaining) }
    private var accent: Color { state.color }
    private var label: String {
        if daysRemaining == 0 { return "HOY" }
        if daysRemaining == 1 || daysRemaining == -1 { return "DÍA" }
        return "DÍAS"
    }

    var body: some View {
        HStack(alignment: .center, spacing: 12) {
            Text(categoria.emoji)
                .font(.system(size: 22))
                .frame(width: 44, height: 44)
                .background(categoryIconBg(categoria),
                            in: RoundedRectangle(cornerRadius: FridgeyRadius.emojiIcon))

            VStack(alignment: .leading, spacing: 2) {
                Text(name)
                    .font(.custom("Inter-Regular", size: 15).weight(.medium))
                    .foregroundStyle(Color.fridgeyInk)
                Text(supporting)
                    .font(.custom("Inter-Regular", size: 12))
                    .foregroundStyle(Color.fridgeyInkMuted)
            }

            Spacer()

            VStack(alignment: .trailing, spacing: 0) {
                if daysRemaining != 0 {
                    Text(String(abs(daysRemaining)))
                        .font(.custom("InstrumentSerif-Regular", size: 26))
                        .foregroundStyle(accent)
                }
                Text(label)
                    .font(.custom("Inter-Regular", size: 10).weight(.medium))
                    .tracking(0.5)
                    .foregroundStyle(accent)
            }
            .frame(minWidth: 64, alignment: .trailing)
        }
        .padding(.horizontal, 18)
        .padding(.vertical, 14)
        .frame(maxWidth: .infinity)
        .overlay(alignment: .bottom) {
            GeometryReader { geo in
                Rectangle()
                    .fill(accent)
                    .frame(width: geo.size.width * progressFraction(daysRemaining), height: 2)
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
            }
        }
    }
}

/// Inverse of days-remaining → bar width fraction (closer to expiry ⇒ wider).
/// Linear over a 14-day horizon; floored at 0.06 so fresh items still show a hint.
private func progressFraction(_ dias: Int) -> CGFloat {
    let horizon: CGFloat = 14
    let clamped = CGFloat(max(0, min(dias, 14)))
    return max(0.06, min(1, 1 - clamped / horizon))
}

File: composeApp/src/androidMain/kotlin/ule/jescuj00/fridgey/ui/components/ProductRow.kt
package ule.jescuj00.fridgey.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.abs
import ule.jescuj00.fridgey.domain.model.Categoria
import ule.jescuj00.fridgey.ui.theme.Amber
import ule.jescuj00.fridgey.ui.theme.EmojiIconShape
import ule.jescuj00.fridgey.ui.theme.Ink
import ule.jescuj00.fridgey.ui.theme.InkMuted
import ule.jescuj00.fridgey.ui.theme.InstrumentSerif
import ule.jescuj00.fridgey.ui.theme.Inter
import ule.jescuj00.fridgey.ui.theme.MintDeep
import ule.jescuj00.fridgey.ui.theme.Rust
import ule.jescuj00.fridgey.ui.theme.Smoke

/**
 * Visual urgency bucket. **The cuts live ONLY in [expirationStateOf]** — the
 * single source of truth reused by (1) the detail section grouping, (2) the
 * ProductRow day-number colour, and (3) the progress bar. Do not re-derive
 * these thresholds anywhere else.
 */
enum class ExpirationState { BAD, WARN, FRESH }

/** Design cuts: <=1 día → BAD (rust), 2..7 → WARN (amber), >7 → FRESH (mint). */
fun expirationStateOf(diasRestantes: Int): ExpirationState = when {
    diasRestantes <= 1 -> ExpirationState.BAD
    diasRestantes <= 7 -> ExpirationState.WARN
    else -> ExpirationState.FRESH
}

/** Day-number / progress-bar colour for a bucket (fresh = mint-deep). */
fun ExpirationState.color(): Color = when (this) {
    ExpirationState.BAD -> Rust
    ExpirationState.WARN -> Amber
    ExpirationState.FRESH -> MintDeep
}

private val NameStyle = TextStyle(
    fontFamily = Inter, fontWeight = FontWeight.Medium, fontSize = 15.sp, letterSpacing = (-0.1).sp,
)
private val SupportingStyle = TextStyle(
    fontFamily = Inter, fontWeight = FontWeight.Normal, fontSize = 12.sp, lineHeight = 16.sp,
)
private val DaysStyle = TextStyle(
    fontFamily = InstrumentSerif, fontWeight = FontWeight.Normal, fontSize = 26.sp, letterSpacing = (-0.5).sp,
)
private val DaysLabelStyle = TextStyle(
    fontFamily = Inter, fontWeight = FontWeight.Medium, fontSize = 10.sp, letterSpacing = 0.5.sp,
)

/**
 * One product inside a "balda". Anatomy per Claude Design:
 *  - 44×44 emoji bubble (radius 14), background tinted per category.
 *  - name (sans 15) + supporting "Categoría · cantidad+unidad" (sans 12, ink-3).
 *  - big serif day count (26) + label (HOY / DÍA / DÍAS), coloured by urgency.
 *  - a 2-px progress bar pinned to the bottom whose width grows as expiry
 *    nears (see [progressFraction]).
 *
 * The urgency [state] is computed from [daysRemaining] via [expirationStateOf]
 * — callers don't pass it, keeping the cuts in one place.
 */
@Composable
fun ProductRow(
    categoria: Categoria,
    name: String,
    supporting: String,
    daysRemaining: Int,
    modifier: Modifier = Modifier,
    onClick: (() -> Unit)? = null,
) {
    val state = expirationStateOf(daysRemaining)
    val accent = state.color()
    val label = when {
        daysRemaining == 0 -> "HOY"
        daysRemaining == 1 || daysRemaining == -1 -> "DÍA"
        else -> "DÍAS"
    }

    val base = if (onClick != null) modifier.fillMaxWidth().clickable(onClick = onClick) else modifier.fillMaxWidth()

    Box(modifier = base) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 18.dp, vertical = 14.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Emoji bubble.
            Box(
                modifier = Modifier
                    .size(44.dp)
                    .clip(EmojiIconShape)
                    .background(categoryIconBg(categoria)),
                contentAlignment = Alignment.Center,
            ) {
                Text(text = categoria.emoji, style = TextStyle(fontSize = 22.sp))
            }

            // Name + supporting.
            Column(
                modifier = Modifier
                    .weight(1f)
                    .padding(horizontal = 12.dp),
            ) {
                Text(text = name, style = NameStyle, color = Ink)
                Spacer(Modifier.height(2.dp))
                Text(text = supporting, style = SupportingStyle, color = InkMuted)
            }

            // Days + label.
            Column(
                modifier = Modifier.widthIn(min = 64.dp),
                horizontalAlignment = Alignment.End,
            ) {
                if (daysRemaining != 0) {
                    Text(text = abs(daysRemaining).toString(), style = DaysStyle, color = accent)
                }
                Text(text = label, style = DaysLabelStyle, color = accent, textAlign = TextAlign.End)
            }
        }

        // Progress bar pinned to the bottom — wider as expiry nears.
        Box(
            modifier = Modifier
                .align(Alignment.BottomStart)
                .fillMaxWidth(progressFraction(daysRemaining))
                .height(2.dp)
                .background(accent),
        )
    }
}

/**
 * Per-category tint for the emoji bubble (design palette). Unlisted categories
 * fall back to smoke.
 */
private fun categoryIconBg(categoria: Categoria): Color = when (categoria) {
    Categoria.LACTEOS -> Color(0xFFEEF2F6)   // dairy
    Categoria.CARNES -> Color(0xFFF5E5E2)     // meat
    Categoria.VERDURAS -> Color(0xFFE5EFE0)   // veg
    Categoria.FRUTAS -> Color(0xFFF6EBDC)     // fruit
    Categoria.BEBIDAS -> Color(0xFFE8EFF1)    // drink
    else -> Smoke                              // other
}

/**
 * Inverse of days-remaining → bar width fraction. Closer to expiry ⇒ wider.
 * Linear over a 14-day horizon: día 0 (or expired) → full width; día 14 →
 * minimal sliver. Floored at 0.06 so even fresh items show a hint of bar.
 */
private fun progressFraction(diasRestantes: Int): Float {
    val horizon = 14f
    val clamped = diasRestantes.coerceIn(0, horizon.toInt())
    return (1f - clamped / horizon).coerceIn(0.06f, 1f)
}

@Preview(showBackground = true, backgroundColor = 0xFFFFFFFF)
@Composable
private fun ProductRowPreview() {
    Column {
        ProductRow(Categoria.VERDURAS, "Espinacas baby", "Verduras · 200 g", 0)
        ProductRow(Categoria.LACTEOS, "Yogur natural", "Lácteos · 4 uds", 3)
        ProductRow(Categoria.FRUTAS, "Manzanas", "Frutas · 6 uds", 12)
    }
}
